distributed/recreate_exceptions.py

```diff
@@ -0,0 +1,171 @@
+            function, args, kwargs = _deserialize(**task)
+            raise gen.Return((function, args, kwargs, deps))
+        else:
+            function, args, kwargs = _deserialize(task=task)
+            raise gen.Return((function, args, kwargs, []))
+
+    def get_futures_error(self, future):
+        """
+        Ask the scheduler details of the sub-task of the given failed future
+
+        When a future evaluates to a status of "error", i.e., an exception
+        was raised in a task within its graph, we an get information from
+        the scheduler. This function gets the details of the specific task
+        that raised the exception and led to the error, but does not fetch
+        data from the cluster or execute the function.
+
+        Parameters
+        ----------
+        future : future that failed, having ``status=="error"``, typically
+            after an attempt to ``gather()`` shows a stack-stace.
+
+        Returns
+        -------
+
+        Tuple:
+        - the function that raised an exception
+        - argument list (a tuple), may include values and keys
+        - keyword arguments (a dictionary), may include values and keys
+        - list of keys that the function requires to be fetched to run
+
+        See Also
+        --------
+        ReplayExceptionClient.recreate_error_locally
+        """
+        return sync(self.client.loop, self._get_futures_error, future)
+
+    @gen.coroutine
+    def _recreate_error_locally(self, future):
+        out = yield self._get_futures_error(future)
+        function, args, kwargs, deps = out
+        futures = self.client._graph_to_futures({}, deps)
+        data = yield self.client._gather(futures)
+        args = pack_data(args, data)
+        kwargs = pack_data(kwargs, data)
+        raise gen.Return((function, args, kwargs))
+
+    def recreate_error_locally(self, future):
+        """
+        For a failed calculation, perform the blamed task locally for debugging.
+
+        This operation should be performed after a future (result of ``gather``,
+        ``compute``, etc) comes back with a status of "error", if the stack-
+        trace is not informative enough to diagnose the problem. The specific
+        task (part of the graph pointing to the future) responsible for the
+        error will be fetched from the scheduler, together with the values of
+        its inputs. The function will then be executed, so that ``pdb`` can
+        be used for debugging.
+
+        Examples
+        --------
+        >>> x0 = delayed(dec)(2)         # doctest: +SKIP
+        >>> y0 = delayed(dec)(1)         # doctest: +SKIP
+        >>> x = delayed(div)(1, x0)      # doctest: +SKIP
+        >>> y = delayed(div)(1, y0)      # doctest: +SKIP
+        >>> tot = delayed(sum)(x, y)     # doctest: +SKIP
+        >>> f = c.compute(tot)           # doctest: +SKIP
+        >>> wait(f)                      # doctest: +SKIP
+        >>> #  f's status becomes "error"
```

Comment:
Perhaps the following:

```python
>>> future = c.submit(div, 1, 0)
>>> future.status
'error'
>>> c.recreate_error_locally(future)
ZeroDivisionError: division by zero

If you're in IPython you might take this opportunity to use pdb

>>> %pdb
Automatic pdb calling has been turned ON

>>> c.recreate_error_locally(future)
ZeroDivisionError: division by zero
      1 def div(x, y):
----> 2     return x / y
ipdb> 
```

Themes of changes:

1.  We start with the simplest possible example, rather than expecting users to read much
2.  We use input/output pairs rather than comments (such as with `>>> future.status` and `ZeroDivisionError`)
3.  I've never referred to pdb in an example before, but in this case it seems like the main point.